jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ProjectAccountSettingsManager2.kt

```diff
@@ -0,0 +1,167 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.core.credentials
+
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.components.ServiceManager
+import com.intellij.openapi.project.Project
+import com.intellij.util.messages.Topic
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import software.aws.toolkits.core.credentials.CredentialProviderNotFound
+import software.aws.toolkits.core.credentials.ToolkitCredentialsProvider
+import software.aws.toolkits.core.region.AwsRegion
+import software.aws.toolkits.jetbrains.core.AwsResourceCache
+import software.aws.toolkits.jetbrains.services.sts.StsResources
+import software.aws.toolkits.jetbrains.utils.MRUList
+import java.util.concurrent.CancellationException
+import kotlin.properties.ObservableProperty
+import kotlin.reflect.KProperty
+
+abstract class ProjectAccountSettingsManager2(private val project: Project) {
+    private val resourceCache = AwsResourceCache.getInstance(project)
+
+    @Transient
+    private var validationJob: Job? = null
+    @Transient
+    private var state: ConnectionState = ConnectionState.INITIALIZING
+
+    private val recentlyUsedProfiles = MRUList<ToolkitCredentialsProvider>(MAX_HISTORY)
+    private val recentlyUsedRegions = MRUList<AwsRegion>(MAX_HISTORY)
+
+    /**
+     * Setting the active region will add to the recently used list, and evict the least recently used if at max size
+     */
+    private var activeRegionInternal: AwsRegion? by ValidateAndNotify(null, recentlyUsedRegions)
+
+    /**
+     * Setting the active provider will add to the recently used list, and evict the least recently used if at max size
+     */
+    private var activeCredentialProviderInternal: ToolkitCredentialsProvider? by ValidateAndNotify(null, recentlyUsedProfiles)
+
+    var activeRegion: AwsRegion
+        get() = activeRegionInternal!! // TODO
+        set(value) {
+            activeRegionInternal = value
+        }
+
+    /**
+     * Setting the active provider will add to the recently used list, and evict the least recently used if at max size
+     */
+    var activeCredentialProvider: ToolkitCredentialsProvider
```

Comment:
I am tempted to just allow this and region to be nullable, it already acts strange where if you have no settings active creds, we will pre-populate all region comboboxes.

I have not done a full check of the blast radius of that change though

Reply:
If we're doing this refactor, I'd like to collapse 'region' and 'credentials' into a single object `AwsConnectionSettings` (or some-such) - we almost *always* use the two together to form a connection to AWS - I don't think it makes sense to have them separate.

Reply:
That is a good call out